src/cargo/core/registry.rs

```diff
@@ -335,35 +357,58 @@ http://doc.crates.io/specifying-dependencies.html#overriding-dependencies
 }
 
 impl<'cfg> Registry for PackageRegistry<'cfg> {
-    fn query(&mut self, dep: &Dependency) -> CargoResult<Vec<Summary>> {
+    fn query(&mut self,
+             dep: &Dependency,
+             f: &mut FnMut(Summary)) -> CargoResult<()> {
         // Ensure the requested source_id is loaded
         self.ensure_loaded(dep.source_id(), Kind::Normal).chain_err(|| {
             format!("failed to load source for a dependency \
                      on `{}`", dep.name())
         })?;
 
+        // Look for an override and get ready to query the real source
         let override_summary = self.query_overrides(&dep)?;
-        let real_summaries = match self.sources.get_mut(dep.source_id()) {
-            Some(src) => Some(src.query(&dep)?),
-            None => None,
+        let mut sources = self.sources.borrow_mut();
```

Comment:
So this is the only place where we need `RefCell`, and it is because we use `self.lock` down the line. `self.lock` only looks at `self.locked` field, so perhaps we can extract `fn lock_helper(locked: &HashMap<to many angle brackets>, summary: Summary) -> Summary` and call it directly in the closure? That way we should please the borrowchecker without RefCell`ing.